### ISSUE TITLE
Agda still blocks on instantiated meta

### ISSUE THREAD

**Issue by martinescardo** [agda/agda#6758]
@andreasabel writes: The OP is fixed by turning off the `__IMPOSSIBLE__` raised for instantiated metas in blockers.
TODO: reinstall the `__IMPOSSIBLE__`, investigate why it is triggered and extract a MWE.

---

OP by @martinescardo :

Agda 2.6.4 RC1 with `TypeTopology` (https://github.com/martinescardo/TypeTopology) gives `__IMPOSSIBLE__` in the module `source/Circle/Integers-SymmetricInduction.lagda`.

```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Constraints.hs:71:11 in Agd-2.6.3.20230805-d3f0739a:Agda.TypeChecking.Constraints
```

**Comment by andreasabel** [comment]
The internal error is here: https://github.com/agda/agda/blob/8c4f28aa99b5b6768206d629402843f4d6c48384/src/full/Agda/TypeChecking/Constraints.hs#L66-L71

**Comment by andreasabel** [comment]
536b73f65801362f0217573b28f106af29014829 is the first bad commit (ping @jespercockx )
> Date:   Fri Nov 11 11:34:05 2022 +0000
> [ re #6211 ] Pull off the band-aid

Maybe too early, the wound is still bleeding...

This issue can be hot-fixed by commenting out the above `__IMPOSSIBLE__` again.

**Comment by jespercockx** [comment]
@martinescardo If possible it would be great to have a standalone example (not depending on any libraries) that still produces the error so I can investigate.

**Comment by martinescardo** [comment]
It is difficult to even begin to think of a standalone example when all the information we get is `__IMPOSSIBLE__`.

**Comment by jespercockx** [comment]
I was investigating this one but it's rather mysterious: it looks like there is a term in `compareAtom` that is blocked on a metavariable, even though *that metavariable does not even occur in the term*. I'm not sure how this is possible, but I'll investigate further tomorrow.

**Comment by jespercockx** [comment]
> This issue can be hot-fixed by commenting out the above __IMPOSSIBLE__ again.

This `__IMPOSSIBLE__` has actually proven to be very valuable since it has already helped to uncover quite a few bugs (both old and new) in the way we manage constraints. Hence I'd rather keep it in the released version, since if anyone runs into it that's actually very valuable information for the development of Agda.

**Comment by jespercockx** [comment]
I found some time to minimize the example today, here is the result:
```agda
_×_ : Set → Set → Set
X × Y = Σ X λ x → Y

is-equiv : (X Y : Set) → (X → Y) → Set
is-equiv X Y f = (Σ (Y → X) λ s → ((x : Y) → f (s x) ≡ x)) × (Σ (Y → X) λ r → ((x : X) → r (f x) ≡ x))

_≃_ : Set → Set → Set
X ≃ Y = Σ (X → Y) λ f → is-equiv X Y f

qinv : (X Y : Set) → (X → Y) → Set
qinv X Y f = Σ (Y → X) λ g → ((x : X) → g (f x) ≡ x) × ((x : Y) → f (g x) ≡ x)

postulate
  qinveq : (X : Set) (Y : Set) (f : X → Y) → qinv X Y f → X ≃ Y

γ = qinveq _ _ _ (ψ , _)
  where
    ψ : _
    ψ _ = _
    η : (x : _) → ψ ≡ x
    η _ = refl
```
Error message:
```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Constraints.hs:71:11 in Agda-2.6.4-C65G09CTfDJ2AVKM71eGU7:Agda.TypeChecking.Constraints
```

**Comment by martinescardo** [comment]
Thanks!

It is still a mystery to me how this situation identifying the problem occurred in only one line out of 165k lines of Agda code.

**Comment by andreasabel** [comment]
> It is still a mystery to me how this situation identifying the problem occurred in only one line out of 165k lines of Agda code.

Agda's source is around the same size.  An immense space of execution pathes, and only a fraction has been taken yet...

**Comment by andreasabel** [comment]
Great, thanks for the fix @jespercockx!  

So you intend to leave the `__IMPOSSIBLE__` in for the release?
Or should we just keep it in the RCs and on `master` but not in the released 2.6.4?

**Comment by jespercockx** [comment]
I removed the impossible just after the release of 2.6.3 with the intention to get rid of all the places that trigger it before the release of 2.6.4. I am *pretty sure* I got all of them by now but as this issue shows it is very hard to be 100% sure of that. I personally prefer to leave it in the released version, but if you don't want to risk it you can also take it out again.

**Comment by martinescardo** [comment]
I prefer to get `__IMPOSSIBLE__` than bad code accepted by accident.

By the way, I take the opportunity to ask this: from time to time I get `__IMPOSSIBLE_` in Agda 2.6.3 when I am trying to fill a hole, but then if I reload the buffer this doesn't occur again.

**Comment by martinescardo** [comment]
Thanks for fixing this issue!

**Comment by jespercockx** [comment]
> I prefer to get __IMPOSSIBLE__ than bad code accepted by accident.

This isn't really a case where unsoundness is a risk, but rather the problem is that without the IMPOSSIBLE we might end up postponing a constraint that will never be tried again. So you'd end up with an "unsolved constraints" error that mentions a constraint that is actually solvable.

**Comment by martinescardo** [comment]
I've just compiled Agda 2.6.4 RC2. I am getting `__IMPOSSIBLE__` again:

```
   Checking Naturals.UniversalProperty (/Volumes/CaseSensitive/feathers/tudo/alive/TypeTopology/source/Naturals/UniversalProperty.lagda).
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Constraints.hs:71:11 in Agd-2.6.3.20230805-d3f0739a:Agda.TypeChecking.Constraints
```
This is in the file `source/Naturals/UniversalProperty.lagda`. Will try to investigate further.

**Comment by martinescardo** [comment]
Apologies. I tested with RC1 by accident - so please disregard the above comment. But I found another error that I will report in another issue.
